cloud/k8s/deploy/shim/shim.go

```diff
@@ -0,0 +1,162 @@
+// +build linux
+
+/*
+   Copyright 2021 Kontain Inc
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+ * This is a containerd shim for KRUN. The idea is a layer on top of the 
+ * standard 'runc' that reconfigures it to use the KRUN binary instead of
+ * RUNC'. This would be trivially easy with C++ (or Java) class inheritance,
+ * but alas GO doesn't do that.
+ *
+ * The only new logic is in Create where we setup an option to reconfigure
+ * BinaryName.
+ */
+
+package shim
+
+import (
+	"context"
+	"os"
+
+	"github.com/containerd/typeurl"
+	"github.com/containerd/containerd/runtime/v2/shim"
+	taskAPI "github.com/containerd/containerd/runtime/v2/task"
+	ptypes "github.com/gogo/protobuf/types"
+	runc_opts "github.com/containerd/containerd/runtime/v2/runc/options"
+	runc_v2 "github.com/containerd/containerd/runtime/v2/runc/v2"
+)
+
+var (
+	// check to make sure the *service implements the GRPC API
+	_ = (taskAPI.TaskService)(&service{})
+)
+
+// New returns a new shim service
+func New(ctx context.Context, id string, publisher shim.Publisher, shutdown func()) (shim.Shim, error) {
+	p, err := runc_v2.New(ctx, id, publisher, shutdown)
+	if err != nil {
+		return nil, err
+	}
+	return &service{
+		parent: p,
+	}, nil
+}
+
+type service struct {
+	parent shim.Shim
+}
+
+// StartShim is a binary call that executes a new shim returning the address
+func (s *service) StartShim(ctx context.Context, opts shim.StartOpts) (string, error) {
+	return s.parent.StartShim(ctx, opts)
+}
+
+// Cleanup is a binary call that cleans up any resources used by the shim when the service crashes
+func (s *service) Cleanup(ctx context.Context) (*taskAPI.DeleteResponse, error) {
+	return s.parent.Cleanup(ctx)
+}
+
+// Create a new container
+func (s *service) Create(ctx context.Context, r *taskAPI.CreateTaskRequest) (_ *taskAPI.CreateTaskResponse, err error) {
+
+	r.Options, err = typeurl.MarshalAny(&runc_opts.Options{
+		BinaryName: "krun",
```

Comment:
How does containerd know where to find krun?
Don't we need to supply an absolute path to it somewhere?